python/ray/serve/router.py

```diff
@@ -146,64 +164,41 @@ def _try_assign_replica(self, query: Query) -> Optional[ray.ObjectRef]:
-        done, _ = ray.wait(refs, num_returns=len(refs), timeout=0)
-        for replica_in_flight_queries in self.in_flight_queries.values():
-            replica_in_flight_queries.difference_update(done)
-        return len(done)
-
-    async def assign_replica(self, query: Query) -> ray.ObjectRef:
-        """Given a query, submit it to a replica and return the object ref.
-        This method will keep track of the in flight queries for each replicas
-        and only send a query to available replicas (determined by the backend
-        max_concurrent_quries value.)
-        """
-        endpoint = query.metadata.endpoint
-        self.num_queued_queries += 1
-        self.num_queued_queries_gauge.set(
-            self.num_queued_queries, tags={"endpoint": endpoint})
-        assigned_ref = self._try_assign_replica(query)
-        while assigned_ref is None:  # Can't assign a replica right now.
-            logger.debug("Failed to assign a replica for "
-                         f"query {query.metadata.request_id}")
-            # Maybe there exists a free replica, we just need to refresh our
-            # query tracker.
-            num_finished = self._drain_completed_object_refs()
-            # All replicas are really busy, wait for a query to complete or the
-            # config to be updated.
-            if num_finished == 0:
-                logger.debug(
-                    "All replicas are busy, waiting for a free replica.")
-                await asyncio.wait(
-                    self._all_query_refs + [self.config_updated_event.wait()],
-                    return_when=asyncio.FIRST_COMPLETED)
-                if self.config_updated_event.is_set():
-                    self.config_updated_event.clear()
-            # We are pretty sure a free replica is ready now, let's recurse and
-            # assign this query a replica.
-            assigned_ref = self._try_assign_replica(query)
-        self.num_queued_queries -= 1
-        self.num_queued_queries_gauge.set(
-            self.num_queued_queries, tags={"endpoint": endpoint})
-        return assigned_ref
+            tracker_ref.future().add_done_callback(
+                functools.partial(
+                    self._on_query_tracker_completed,
+                    replica_handle=replica,
+                    tracker_ref=tracker_ref))
+            # user_ref is written to the concurrent future that the caller of
+            # router gets immediately when enqueued a query.
+            pending_query.ref_future.set_result(user_ref)
+
+            return True
+        return False
+
+    def _on_query_tracker_completed(self, _tracker_result, *,
+                                    replica_handle: ActorHandle,
+                                    tracker_ref: ray.ObjectRef):
+        tracker_refs: Set[ray.ObjectRef] = self.in_flight_queries.get(
+            replica_handle, {})
+        tracker_refs.discard(tracker_ref)
+        self._try_send_pending_queries()
```

Comment:
what's the behavior when the actor request fails? do we need to handle it differently at all?